File: backend/app/models/media.py
from datetime import datetime
from typing import Optional
from uuid import UUID

from sqlalchemy import (
    BigInteger,
    CheckConstraint,
    DateTime,
    ForeignKey,
    Integer,
    Text,
    UniqueConstraint,
    text,
)
from sqlalchemy.orm import Mapped, mapped_column

from app.models.base import Base, created_at_col, uuid_pk
from app.models.enums import (
    MEDIA_LAYER,
    MEDIA_STATUS,
    PUBLICATION_STATE,
    MediaLayer,
    MediaStatus,
    PublicationState,
)


class Media(Base):
    """One logical photograph — and, since CK-32, the ingest job that
    produces it (media pipeline record §6.1: the media row IS the job; a
    `media_jobs` table beside `status` would be two places that can disagree
    about whether a photograph exists, the shape CK-13 banned for the
    refcount).

    Provenance (uploader, timestamp, gathering) is NOT NULL by design — the
    book cannot be built without it (roadmap §2). Media hangs off the
    GATHERING because the gathering is the keeping and storage unit (CK-12);
    the optional occurrence_id is a label, never an owner — which is why its
    FK carries ON DELETE SET NULL (0016): removing a date removes the label
    and keeps the photograph. Not a cascade, and not a third confirmable
    refusal beside CK-30's two.

    THIS ROW STORES NO OBJECT. The stored representations of the photograph
    are MediaDerivative rows (one per layer, below); the quarantine original
    that exists between `uploaded` and `ready` is addressed by a key DERIVED
    from this row's id (the intent endpoint and the worker compute the same
    key) and is never written to a column — a stored pointer to an object
    the publish step deletes would be a second representation of "is there
    still an original", beside `status`, waiting to disagree with it. The two
    `upload_*` columns are facts about the UPLOAD, fixed at intent and true
    at every rung: the declared MIME type (the allowlist's input) and the
    declared byte size (the presigned PUT signs it as Content-Length; the
    quota reservation sums it over an account's pending rows — record §6.6).
    Neither is the photograph's stored size or format — those are sums and
    values over the derivative rows, which is why the 0001 names
    `size_bytes`/`content_type` were retired rather than reinterpreted.

    Two gates, two columns, never collapsed (record §5): `status` is the
    machine's answer (the MediaStatus ladder) and `publication_state` is the
    host's. `status` has NO server default — a writer must state the rung:
    a default of `processing` would let an omitted status be reclaimed by
    the worker fifteen minutes later against an object that was never
    uploaded.

    Job bookkeeping (record §6.2) rides here as columns: `attempts` (retry
    count), `available_at` (not claimable before — the backoff ladder),
    `claimed_at` (the reclaim clock), `last_error` (dead-letter reason; names
    the outcome, never the file's contents)."""

    __tablename__ = "media"
    __table_args__ = (
        CheckConstraint(
            "uploader_person_id IS NOT NULL OR guest_name IS NOT NULL", name="identity_present"
        ),
    )

    id: Mapped[UUID] = uuid_pk()
    gathering_id: Mapped[UUID] = mapped_column(
        ForeignKey("gatherings.id"), nullable=False, index=True
    )
    # A label, never an owner: the photograph survives the removal of its
    # date and loses only the label (0016 — the CK-30 dormant 500, closed
    # for media by the phase that gives it a surface).
    occurrence_id: Mapped[Optional[UUID]] = mapped_column(
        ForeignKey("occurrences.id", ondelete="SET NULL"), nullable=True, index=True
    )
    uploader_person_id: Mapped[Optional[UUID]] = mapped_column(
        ForeignKey("people.id"), nullable=True
    )
    guest_name: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    # Decided at CK-34 (0017): NULL while the row waits in `pending_upload`,
    # stamped by the confirm step in the same transaction that moves status
    # to `uploaded`, never touched again — claimed_at's shape. The row is
    # born at intent, and that instant is already `created_at`; 0016's
    # now() default stamped intent time under this name, and an `intent_at`
    # rename would have been a second column carrying created_at's value.
    uploaded_at: Mapped[Optional[datetime]] = mapped_column(
        DateTime(timezone=True), nullable=True
    )
    # Facts about the upload, fixed at intent (see the class docstring).
    upload_content_type: Mapped[str] = mapped_column(Text, nullable=False)
    upload_size_bytes: Mapped[int] = mapped_column(BigInteger, nullable=False)
    # The machine's gate. No server default, deliberately.
    status: Mapped[MediaStatus] = mapped_column(MEDIA_STATUS, nullable=False)
    # Job bookkeeping (record §6.2). attempts counts from zero; available_at
    # is stamped by the confirm step (the claim predicate is
    # `available_at <= now()`, so a confirmed row must carry one — CK-34);
    # claimed_at and last_error are NULL until the worker first touches
    # the row.
    attempts: Mapped[int] = mapped_column(
        Integer, nullable=False, server_default=text("0")
    )
    available_at: Mapped[Optional[datetime]] = mapped_column(
        DateTime(timezone=True), nullable=True
    )
    claimed_at: Mapped[Optional[datetime]] = mapped_column(
        DateTime(timezone=True), nullable=True
    )
    last_error: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    # The host's gate — the ONE publication state the three layers share
    # (media-layers record §8). Revocation and removal act here, on the
    # logical photograph; the layers have no state of their own.
    publication_state: Mapped[PublicationState] = mapped_column(
        PUBLICATION_STATE, nullable=False
    )
    removed_at: Mapped[Optional[datetime]] = mapped_column(
        DateTime(timezone=True), nullable=True
    )
    # Words on the photograph (0018, CK-39;
    # decisions/2026-09-10-captions-tags-and-finding-a-photograph.md).
    # `filename` is the uploader's original file name, taken at intent and
    # never changed — A DISPLAY STRING AND NOTHING ELSE: it never reaches a
    # storage key (quarantine_key / published_key are pure functions of the
    # row id), a path, or a header, and nothing relies on its extension (the
    # declared content type is advisory; the worker's decode decides).
    # Nullable with no backfill: every row written before 0018 reads NULL,
    # and NULL is the one representation of "no name" — a blank is refused
    # at the intent (the CK-13/CK-20 discipline). `caption` is the person's
    # own words over the file name, set by the uploader alone through
    # PATCH /media/{id} under merge-patch semantics (CK-22): explicit null
    # clears it, a blank is refused, absent leaves it alone. Tags are rows
    # (MediaTag, below), never an array column here.
    filename: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    caption: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    created_at: Mapped[datetime] = created_at_col()


class MediaTag(Base):
    """CK-39 — one free-text tag on one photograph: the UPLOADER'S OWN WORDS
    about their own photograph, and nothing more (0018;
    decisions/2026-09-10-captions-tags-and-finding-a-photograph.md).

    A table, not an array column on media: gathering-scoped tags and person
    tags are both plausible later, and a table makes each a smaller change
    (a decision already made before this phase; not re-opened here). One
    row per (photograph, tag) — `UNIQUE (media_id, tag)` — so a photograph
    cannot carry the same tag twice; the API also refuses two tags that
    differ only in case, treating them as one (case is preserved as typed,
    equality is case-insensitive — the search is too).

    THIS TABLE STORES NO GATHERING AND NO PERSON. No `gathering_id`: a tag
    hangs off a media row, which hangs off a gathering, and a column here
    would be a second representation of a fact the join already answers —
    the defect CK-13 banned for the refcount and CK-32 kept out of
    media_derivatives. And NO `person_id` for a tagged person, not even a
    nullable one: a free-text tag is a claim the uploader makes about their
    own photograph; a PERSON tag is a claim about someone else — the shape
    consent-and-compliance.md carries an open finding on (companion names,
    CK-29: personal data about a third party who never entered it and
    holds no account). When person tagging is decided it gets its own
    record and its own table, with the consent questions answered rather
    than inherited; a nullable `person_id` here would smuggle the feature
    in as a schema detail. `added_by_person_id` is provenance — who wrote
    the words — never a subject; it does not cascade from the person
    (account deletion is anonymization; the row stays attributed to the
    anonymized person, like every contribution).

    Rows are replaced wholesale on every tag write (the list IS the value —
    the companions shape) and deleted with their photograph
    (`ON DELETE CASCADE`): a tag on a deleted photograph is an assertion
    with nothing behind it, and a takedown removes the words with the
    picture, so no assertion outlives its subject."""

    __tablename__ = "media_tags"
    __table_args__ = (
        # One photograph, one tag, once. Leads with media_id, so it also
        # indexes the FK — no separate index (the 0014/0016 pattern).
        UniqueConstraint("media_id", "tag"),
    )

    id: Mapped[UUID] = uuid_pk()
    media_id: Mapped[UUID] = mapped_column(
        ForeignKey("media.id", ondelete="CASCADE"), nullable=False
    )
    tag: Mapped[str] = mapped_column(Text, nullable=False)
    # Provenance, never a subject (see the class docstring).
    added_by_person_id: Mapped[UUID] = mapped_column(
        ForeignKey("people.id"), nullable=False
    )
    created_at: Mapped[datetime] = created_at_col()


class MediaDerivative(Base):
    """One stored object: one layer (archival / web / thumbnail) of one
    photograph (CK-32; decisions/2026-09-07-media-derivative-schema-shape.md).
    The layers differ in storage class, in size, and in access pattern
    (`last_accessed` matters to the archival layer's tiering and never to
    the thumbnail's), which is what makes them rows rather than nine
    columns on media — and it is what lets the format change per layer
    (WebP today, AVIF later — media-layers §1/§5) without a migration, since
    `content_type` and `storage_key` record what was actually written.

    THE BINDING (media-layers record §8; consent record §1): this table has
    NO publication_state, NO removed_at, NO status, and no lifecycle column
    of its own — ever. Every piece of state lives on the media row, so "the
    three layers die as a unit" is enforced by there being no per-layer
    state to diverge, rather than remembered by every writer. The FK
    cascades for the same reason: a derivative row cannot outlive its
    photograph (the 0014 companions shape — containment, not a content
    cascade; nothing here cascades from a person). `verify_schema.py`
    asserts the absent columns so a later phase cannot quietly add one.

    Rows are written by the worker in the same transaction that moves the
    media row to `ready` — publish is the last step and is atomic (pipeline
    record §8) — so a derivative row never exists under a row that is not
    ready; the verifier asserts that too."""

    __tablename__ = "media_derivatives"
    __table_args__ = (
        # One row per layer per photograph. Leads with media_id, so it also
        # indexes the FK — no separate index (the 0014 companions pattern).
        UniqueConstraint("media_id", "layer"),
    )

    id: Mapped[UUID] = uuid_pk()
    media_id: Mapped[UUID] = mapped_column(
        ForeignKey("media.id", ondelete="CASCADE"), nullable=False
    )
    layer: Mapped[MediaLayer] = mapped_column(MEDIA_LAYER, nullable=False)
    # The published-bucket object key. One object, one row.
    storage_key: Mapped[str] = mapped_column(Text, nullable=False, unique=True)
    # The layer's own MIME type — recorded, not derived from the layer,
    # because the web layer's format is expected to change.
    content_type: Mapped[str] = mapped_column(Text, nullable=False)
    # Bytes actually stored. The sum over a photograph's rows is what moves
    # gatherings.total_bytes at publish (record §6.6).
    size_bytes: Mapped[int] = mapped_column(BigInteger, nullable=False)
    # Open text (database-schema decision 2 — R2 class names may change).
    # No server default: the worker states the class per layer; a default
    # of 'standard' would let an archival row silently land on the wrong
    # class.
    storage_class: Mapped[str] = mapped_column(Text, nullable=False)
    # Tiering input (~Phase 5); per object, which is the reason it moved
    # off the media row.
    last_accessed: Mapped[Optional[datetime]] = mapped_column(
        DateTime(timezone=True), nullable=True
    )
    created_at: Mapped[datetime] = created_at_col()

File: frontend/src/lib/media.ts
// Media API types and the upload/read helpers (CK-38). Shapes mirror
// backend/app/api/media.py's response bodies exactly — see
// reference/backend/api-reference.md, Media router.
//
// THE ONE RULE THIS MODULE EXISTS TO KEEP (pipeline record §9): a presigned
// URL — the PUT the intent endpoint issues, the GET the url endpoint mints —
// is a bearer credential carrying an Access Key ID and a valid signature.
// It is spent on exactly one fetch and exists nowhere else: never in React
// state, never in a log, never in an error message, never in anything a
// retry could read back. The two helpers below take the response body as a
// local, make the one call, and return something that is not the URL — a
// verdict for the PUT, an object URL for the read (a `blob:` reference to
// bytes already in this browser, which carries no credential at all).
//
// THE WORDS ON A PHOTOGRAPH (CK-40, the surface for CK-39's backend;
// decisions/2026-09-10-captions-tags-and-finding-a-photograph.md): a row is
// called by its caption, then its filename, then "Photo" (mediaName); the
// caption follows the Patch-semantics convention (captionPatch — blank is
// never a clear, explicit null is); and the tag list REPLACES the set on the
// server, so tagsPatch always carries every tag the photograph should end up
// with — sending only the new one silently deletes the rest.
//
// THE HOST'S REVIEW (CK-43.1, the surface for CK-43's backend;
// decisions/2026-09-13-the-hosts-review.md): the queue is the same list
// asked for with `awaiting_review=true` (mediaListPath); publish and decline
// are the two acts (reviewMedia), each refused with a STABLE CODE the copy
// switches on and never the server's wording (reviewRefusalMessage); and the
// pending line becomes conditional on the gathering resolving gated
// (mediaStateMessage's `gated` — record §8: where nobody reviews, the line
// says who can see the photograph and nothing about what it waits on).
import { authFetch } from './api'
import { networkErrors } from './formErrors'

export type MediaStatus = 'pending_upload' | 'uploaded' | 'processing' | 'ready' | 'failed'
export type PublicationState = 'pending' | 'live' | 'removed'
export type ServableLayer = 'web' | 'thumbnail'

// A GET /gatherings/{id}/media row: the CK-34 media body plus who uploaded
// it (a display name — never an email, never a person id), whether it is
// the caller's own, and the bin clock. There is NO url field, by design.
export interface MediaItem {
  id: string
  gathering_id: string
  occurrence_id: string | null
  // Typed loosely on purpose: a rung or state this build does not know is
  // rendered honestly by name (see mediaStateMessage), never as a blank.
  status: MediaStatus | string
  publication_state: PublicationState | string
  upload_content_type: string
  upload_size_bytes: number
  uploaded_at: string | null
  created_at: string
  uploader_display_name: string | null
  is_own: boolean
  removed_at: string | null
  // The words (CK-39): the file's name as taken at intent — a display string
  // and nothing else, NULL on every row written before migration 0018 and
  // never backfilled; the uploader's caption, NULL when there is none (never
  // ""); and the tags, case-insensitively alphabetical — a set, not the
  // order typed.
  filename: string | null
  caption: string | null
  tags: string[]
  // The publication stamp's date (CK-43; migration 0020): null until
  // published, set by the worker where the gathering resolves open and by
  // the host's publish where it is gated. Carried because the body carries
  // it and RENDERED NOWHERE — the stamp is consent evidence, never row
  // furniture (record §7), and the publisher's person id rides no body.
  published_at: string | null
}

export interface MediaList {
  media: MediaItem[]
}

// One entry of the intent response: the new row and the credential for ONE
// PUT of exactly the declared bytes. `headers` carries the values R2 signed
// in — the PUT must send Content-Type exactly as given; Content-Length is
// the browser's to set from the body (a forbidden header in fetch).
export interface UploadInstruction {
  url: string
  method: string
  headers: Record<string, string>
  expires_in: number
}

export interface UploadIntent {
  // The intent body carries the filename it stored (trimmed) and never the
  // tags — a body that did not load them must not claim [].
  media: { id: string; status: string; filename: string | null }
  upload: UploadInstruction
}

export interface IntentResponse {
  intents: UploadIntent[]
}

// The server's allowlist, mirrored for the picker's `accept` — and NEVER
// the authority: a `.mov` renamed `.jpg` passes any client check, the
// server's 422 is what refuses it, and CK-36's worker is what catches a
// lie the 422 cannot see. Extensions are listed as well as types because a
// HEIC's `file.type` is empty in the browser (observed at CK-36's (da)) and
// the picker filters by whichever the platform knows.
export const ACCEPTED_CONTENT_TYPES = [
  'image/jpeg',
  'image/png',
  'image/webp',
  'image/heic',
  'image/heif',
]
export const PICKER_ACCEPT = [
  ...ACCEPTED_CONTENT_TYPES,
  '.jpg',
  '.jpeg',
  '.png',
  '.webp',
  '.heic',
  '.heif',
].join(',')

// How often the list re-reads while a photograph is still moving (CK-36
// measured claim-to-ready at ~4.7 s, so a few seconds is the right grain);
// a page with nothing in flight polls nothing at all.
export const POLL_INTERVAL_MS = 3000

// The bin's read window (api/media.py REMOVED_BIN) — stated in copy only.
export const REMOVED_BIN_DAYS = 30

const TYPE_BY_EXTENSION: Record<string, string> = {
  jpg: 'image/jpeg',
  jpeg: 'image/jpeg',
  png: 'image/png',
  webp: 'image/webp',
  heic: 'image/heic',
  heif: 'image/heif',
}

// The content type the intent declares. `file.type` when the browser knows
// it; otherwise the extension — a HEIC arrives with an EMPTY type, and the
// declared value is signed into the presigned PUT, so a wrong or missing
// one produces an R2 refusal that reads like a permissions problem. A type
// nobody can name goes out as the generic binary type so the SERVER refuses
// it with its own message — the client never pre-empts the allowlist.
export function declaredContentType(file: { name: string; type: string }): string {
  if (file.type !== '') return file.type
  // The name trimmed (CK-40): a trailing space after the extension would
  // otherwise declare the generic type for a perfectly good HEIC.
  const name = file.name.trim()
  const dot = name.lastIndexOf('.')
  const extension = dot === -1 ? '' : name.slice(dot + 1).toLowerCase()
  return TYPE_BY_EXTENSION[extension] ?? 'application/octet-stream'
}

// The 422 keys the intent form renders inline, per chosen file (the batch is
// refused whole, so every index the caller sent may carry an error) plus the
// batch-level key the quota, ceiling, and count refusals land on.
export function intentErrorFields(count: number): string[] {
  const fields = ['items']
  for (let index = 0; index < count; index++) {
    fields.push(
      `items.${index}.content_type`,
      `items.${index}.size_bytes`,
      `items.${index}.occurrence_id`,
      `items.${index}.filename`,
    )
  }
  return fields
}

// One intent item from one File — and from nothing else. `size_bytes` is
// `file.size` because the presigned PUT SIGNS Content-Length, which fetch
// forbids the client from setting: the browser derives it from the body, and
// the upload validates only while the declared size and the blob agree
// EXACTLY. A size captured anywhere else — a form field, a value recomputed
// after a resize or a strip — makes R2 reject on the signature, which reads
// as a credentials failure nowhere near the real cause (CK-39's (dp)). The
// filename is trimmed before it goes (the server trims too); a name that is
// blank after trimming is left out rather than sent for the server to
// refuse the whole batch over.
export function intentItem(file: File): {
  content_type: string
  size_bytes: number
  filename?: string
} {
  const filename = file.name.trim()
  return {
    content_type: declaredContentType(file),
    size_bytes: file.size,
    ...(filename === '' ? {} : { filename }),
  }
}

// What a row is called: the person's own words win, the file's name is the
// fallback, and "Photo" is what remains when there is neither — which is
// every row uploaded before CK-39, permanently, since nothing is backfilled.
export function mediaName(item: Pick<MediaItem, 'caption' | 'filename'>): string {
  if (item.caption !== null && item.caption !== '') return item.caption
  if (item.filename !== null && item.filename !== '') return item.filename
  return 'Photo'
}

// How long the search box waits after the last keystroke before asking the
// server: a request per keystroke against a substring scan is the wrong
// shape even at this size.
export const SEARCH_DEBOUNCE_MS = 300

// The list path, with `q` only when there is a term. An empty box sends no
// `q` at all — never an empty one — and a whitespace-only entry is the same
// as an empty box (the server treats blank as "not a search" too). Search
// is per gathering: the endpoint is, and so is this.
//
// The host's queue (CK-43.1) is the SAME list asked for with
// `awaiting_review=true` — a filter on the list, never a sibling route
// (record §10), so the surface has one list shape; it composes with `q`
// on the server, inside the audience rule. The flag goes only when the
// queue is what is wanted — never `awaiting_review=false`.
export function mediaListPath(
  gatheringId: string,
  term: string,
  options: { awaitingReview?: boolean } = {},
): string {
  const params: string[] = []
  if (options.awaitingReview) params.push('awaiting_review=true')
  const trimmed = term.trim()
  if (trimmed !== '') params.push(`q=${encodeURIComponent(trimmed)}`)
  const base = `/gatherings/${gatheringId}/media`
  return params.length === 0 ? base : `${base}?${params.join('&')}`
}

// The caption under the Patch-semantics convention (CK-22 — the same four
// lines GatheringDetail's occurrence editor applies to location and map link;
// decisions/2026-08-27-optional-field-clearing.md). `undefined` means leave
// the field out of the patch; `null` means clear; a string is the value to
// store. Three rules: "had a saved caption, now blank" goes out as explicit
// null (never ""); "was empty, still empty" is omitted, never a null no-op
// write; a whitespace-only entry goes out AS TYPED so the server's
// blank-rejection 422 renders inline — the client never collapses whitespace
// into a clear, because blank is never a clear.
export function captionPatch(raw: string, saved: string | null): string | null | undefined {
  const savedText = saved ?? ''
  const trimmed = raw.trim()
  if (raw === '') return savedText === '' ? undefined : null
  if (trimmed === '') return raw
  return trimmed === savedText ? undefined : trimmed
}

// The tag list to send: EVERY tag the photograph should end up with, or
// nothing when the set is unchanged. The server REPLACES the set with what
// it receives — a patch carrying only the new tag would delete the others
// with no error and no warning, data loss that reads as a UI bug. So the
// editor's whole list goes, on an add and on a removal alike; `[]` is how
// the last tag is removed. Compared as sets of the exact strings (the
// server preserves case as typed, so a case change is a real edit; a tag
// removed and re-added is not).
export function tagsPatch(tags: string[], saved: string[]): string[] | undefined {
  const next = [...tags].sort()
  const current = [...saved].sort()
  const same = next.length === current.length && next.every((tag, index) => tag === current[index])
  return same ? undefined : tags
}

export interface MediaWords {
  caption: string
  tags: string[]
}

// The PATCH /media/{id} body for the editor's state against the saved row —
// merge-patch semantics: a field absent leaves the stored value alone. An
// empty object means nothing changed and no request should go.
export function mediaWordsPatch(
  form: MediaWords,
  saved: Pick<MediaItem, 'caption' | 'tags'>,
): { caption?: string | null; tags?: string[] } {
  const patch: { caption?: string | null; tags?: string[] } = {}
  const caption = captionPatch(form.caption, saved.caption)
  if (caption !== undefined) patch.caption = caption
  const tags = tagsPatch(form.tags, saved.tags)
  if (tags !== undefined) patch.tags = tags
  return patch
}

// The 422 keys the words editor renders inline: the caption, the tag list as
// a whole (the count refusal), and each sent tag by its index — a duplicate
// is refused on the entry that repeats it, and lands on that entry.
export function wordsErrorFields(tagCount: number): string[] {
  const fields = ['caption', 'tags']
  for (let index = 0; index < tagCount; index++) fields.push(`tags.${index}`)
  return fields
}

// `ready` and `failed` are the ladder's two ends; every other rung is a
// photograph still moving, and the list keeps re-reading while one exists.
export function isTerminal(item: Pick<MediaItem, 'status'>): boolean {
  return item.status === 'ready' || item.status === 'failed'
}

export function anyInFlight(items: Pick<MediaItem, 'status'>[]): boolean {
  return items.some((item) => !isTerminal(item))
}

// What the person is told about a row — A FUNCTION OF THE ROW'S STATE, NEVER
// A CONSTANT. `status` is the machine's answer (is it processed?);
// `publication_state` is the human's (who may see it?). Both are read from
// the row, so when the publication phase starts moving rows to `live`
// (decisions/2026-09-09-consent-gate-defaults.md §1: a private, person-owned
// gathering does not wait for anyone) this function changes nothing — the
// `live` branch is written now for a case nothing reaches yet, exactly as
// CK-37 wrote its `live` audience branch.
//
// Two things this copy may never say for a gathering that resolves OPEN
// (consent-gate-defaults §9; the CK-37 record's DATA-HANDLING): that anyone
// will review or has approved the photograph — for a family gathering nobody
// ever will, so that copy is wrong today and wrong after the fix — and that
// a `pending` photograph is shared or visible to the gathering. It is not;
// the audience rule is what makes that true. The `failed` line names the
// outcome, never the file's contents (pipeline record §6.5: "we couldn't
// process this photo", not "corrupt").
//
// Where the gathering resolves GATED (`viewer.gated` — the detail body's
// effective `requires_approval`, CK-41) someone genuinely does decide, and
// the pending line saying nothing about it would be the opposite lie
// (the-hosts-review §8). So, and only then, the line gains a second
// sentence saying what the photograph waits on. It is read by the uploader,
// who may not be the host, and by the host, who in a family gathering
// usually IS the uploader — so the host's copy addresses them as the one
// who decides ("waiting for you"), never as a person waiting on "the host",
// which would be themselves. Both readings are true at once. Nothing here
// names the product as the reviewer: a host publishes or declines.
export function mediaStateMessage(
  item: Pick<MediaItem, 'status' | 'publication_state' | 'is_own' | 'uploader_display_name'>,
  viewer: { isHost: boolean; gated?: boolean },
): string {
  switch (item.status) {
    case 'pending_upload':
      return 'On its way — still being uploaded.'
    case 'uploaded':
      return 'Uploaded — waiting to be prepared.'
    case 'processing':
      return 'Being prepared — usually within a minute, longer for a big batch.'
    case 'failed':
      return "We couldn't process this photo."
    case 'ready':
      break
    default:
      // A rung this build does not know: name it rather than show nothing.
      return `Status: ${item.status}.`
  }

  // Ready: who may see it is decided by publication_state
  // (decisions/2026-09-09-who-may-see-an-unapproved-photograph.md §1).
  switch (item.publication_state) {
    case 'live':
      return 'Everyone in this gathering can see it.'
    case 'pending': {
      // The uploader and the host, and nobody else. Spelled from the
      // viewer's seat: the two admitted people collapse to one when the
      // uploader is the host.
      const uploader = item.uploader_display_name ?? 'the person who added it'
      const audience = item.is_own
        ? viewer.isHost
          ? 'Only you can see this.'
          : 'Only you and the host can see this.'
        : viewer.isHost
          ? `Only ${uploader} and you can see this.`
          : `Only ${uploader} and the host can see this.`
      if (!viewer.gated) return audience
      // Gated: what it waits on, from the same seat. The host decides, so
      // the host is told it waits for them — the common case in a family
      // gathering is a host reading their own photograph.
      return viewer.isHost
        ? `${audience} Waiting for you to publish or decline it.`
        : `${audience} Waiting for the host to publish or decline it.`
    }
    case 'removed':
      return item.is_own
        ? `Removed. Only you can still see it, for ${REMOVED_BIN_DAYS} days after removal.`
        : 'Removed.'
    default:
      return `Ready — ${item.publication_state}.`
  }
}

// ---------------------------------------------------------------------------
// The host's review (CK-43.1). The queue is the list with the flag; the two
// acts are below; the batch lives in the component beside the upload batch.
// ---------------------------------------------------------------------------

// What the host may act on: a `ready` row that is `pending` (record §3) —
// the server's own criterion for the queue, mirrored here ONLY to decide
// whether a row carries the publish and decline controls. It is a predicate
// over the row's own state, never state of its own: an affordance on a row
// the server would refuse (a `live` one, a `failed` one) is its own defect.
export function awaitingReview(item: Pick<MediaItem, 'status' | 'publication_state'>): boolean {
  return item.status === 'ready' && item.publication_state === 'pending'
}

// The batch endpoint takes up to fifty ids (api-reference, the batch form).
export const MAX_PUBLISH_BATCH = 50

export type ReviewAct = 'publish' | 'decline'

export type ReviewOutcome =
  | { ok: true }
  // The 409's stable code (already_live | not_pending | not_ready) with the
  // state the server reports beside it, `not_found` for the 404 (the row
  // moved out of the host's sight — the bin is the uploader's), or this
  // module's own words for the rest. The server's `message` is NOT carried:
  // the codes are the contract, the strings are ours (record §3).
  | { ok: false; code: string; publication_state?: string; status?: string }

// One act on one photograph. A guarded update on the server: a row that
// moved since the list was read is refused with its current state, never
// changed, and the code says which state that is.
export async function reviewMedia(mediaId: string, act: ReviewAct): Promise<ReviewOutcome> {
  try {
    const response = await authFetch(`/media/${mediaId}/${act}`, { method: 'POST' })
    if (response.ok) return { ok: true }
    if (response.status === 404) return { ok: false, code: 'not_found' }
    if (response.status === 409) {
      const body = (await response.json().catch(() => null)) as {
        detail?: { code?: string; publication_state?: string; status?: string }
      } | null
      return {
        ok: false,
        code: body?.detail?.code ?? 'failed',
        publication_state: body?.detail?.publication_state,
        status: body?.detail?.status,
      }
    }
    return { ok: false, code: 'failed' }
  } catch {
    return { ok: false, code: 'unreachable' }
  }
}

// What a refused act tells the host — switched on the CODE and never the
// wording (record §3: the codes are the contract, the strings are ours), so
// a reworded server message changes nothing here. Each line says what state
// the photograph is actually in, and none claims anything was destroyed:
// a declined photograph is in its uploader's bin for thirty days.
export function reviewRefusalMessage(
  act: ReviewAct,
  refusal: { code: string; publication_state?: string; status?: string },
): string {
  switch (refusal.code) {
    case 'already_live':
      return 'This photo is already published — everyone in this gathering can see it.'
    case 'not_pending':
      if (refusal.publication_state === 'live') {
        // Decline on a published photograph: the takedown is unbuilt
        // (record §11), and the copy says so rather than pretending.
        return act === 'publish'
          ? 'This photo is already published — everyone in this gathering can see it.'
          : "This photo is already published, and taking a published photo down isn't possible here yet."
      }
      return act === 'publish'
        ? "This photo was declined or removed, so it can't be published."
        : 'This photo was already declined or removed.'
    case 'not_ready':
      return refusal.status === 'failed'
        ? "This photo couldn't be processed, so there's nothing to publish."
        : "This photo isn't ready yet — it can't be published or declined until it is."
    case 'not_found':
      return "This photo isn't here any more — refresh the list to see what's waiting."
    case 'unreachable':
      return networkErrors().form[0]
    default:
      return 'Something went wrong. Nothing changed — try again.'
  }
}

// The 422 keys the batch form renders inline: the list as a whole (the
// count bounds) and each sent id by its index — the server refuses the
// batch WHOLE on the first bad entry and lands the refusal on that entry
// (`["body","media_ids",N]`), carrying the single act's code beside the
// message, so the one mapper renders it on the row that caused it.
export function publishBatchErrorFields(count: number): string[] {
  const fields = ['media_ids']
  for (let index = 0; index < count; index++) fields.push(`media_ids.${index}`)
  return fields
}

// ---------------------------------------------------------------------------
// The two calls that spend a presigned URL. Nothing below returns one.
// ---------------------------------------------------------------------------

export type PutOutcome = 'stored' | 'refused' | 'unreachable'

// The one PUT. Sends the Content-Type R2 signed (the declared type, exactly
// as the server normalised it) and lets the browser set Content-Length from
// the body — setting it by hand is forbidden in fetch, and the body IS the
// declared size, which is what makes the signature match. No Authorization
// header: this is not our API, and a bearer JWT must never travel to R2.
export async function putUpload(file: Blob, upload: UploadInstruction): Promise<PutOutcome> {
  try {
    const response = await fetch(upload.url, {
      method: upload.method,
      headers: { 'Content-Type': upload.headers['Content-Type'] },
      body: file,
      credentials: 'omit',
    })
    return response.ok ? 'stored' : 'refused'
  } catch {
    // A network failure or a CORS refusal at preflight — the kickoff names
    // the latter as bucket configuration, an out-of-band fix, not a code bug.
    return 'unreachable'
  }
}

export type ConfirmOutcome =
  | { ok: true }
  // The 409's stable code (object_missing | size_mismatch | not_pending), the
  // 503's storage_unavailable, or this module's own words for the rest.
  | { ok: false; code: string }

// Confirm is the server VERIFYING rather than trusting: it HEADs the key and
// moves the row only if the object is there at the declared size.
export async function confirmUpload(mediaId: string): Promise<ConfirmOutcome> {
  try {
    const response = await authFetch(`/media/${mediaId}/confirm`, { method: 'POST' })
    if (response.ok) return { ok: true }
    if (response.status === 409 || response.status === 503) {
      const body = (await response.json().catch(() => null)) as {
        detail?: { code?: string }
      } | null
      return { ok: false, code: body?.detail?.code ?? 'failed' }
    }
    return { ok: false, code: 'failed' }
  } catch {
    return { ok: false, code: 'unreachable' }
  }
}

export type LayerOutcome =
  | { ok: true; objectUrl: string }
  // not_ready: the row is visible but has no object yet (409). unavailable:
  // anything else — including a CORS refusal on the published bucket, which
  // is the bucket's configuration, not this code's.
  | { ok: false; reason: 'not_ready' | 'unavailable' }

// One layer of one ready photograph, as bytes in this browser. Mints the
// presigned GET through our API, spends it on ONE fetch, and hands back a
// `blob:` object URL — which is what goes into an <img> and into state. The
// presigned URL itself is a local of this function and dies with it. The
// caller owns the object URL and revokes it when done (URL.revokeObjectURL).
// Never `archival`: the print master is never served (CK-37 refuses it).
export async function fetchLayerObjectUrl(
  mediaId: string,
  layer: ServableLayer,
): Promise<LayerOutcome> {
  try {
    const minted = await authFetch(`/media/${mediaId}/url?layer=${layer}`)
    if (minted.status === 409) return { ok: false, reason: 'not_ready' }
    if (!minted.ok) return { ok: false, reason: 'unavailable' }
    const body = (await minted.json()) as { url: string; method: string }
    const object = await fetch(body.url, { method: body.method, credentials: 'omit' })
    if (!object.ok) return { ok: false, reason: 'unavailable' }
    return { ok: true, objectUrl: URL.createObjectURL(await object.blob()) }
  } catch {
    return { ok: false, reason: 'unavailable' }
  }
}
